File: backend/internal/protocol/frame.go
package protocol

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
)

// Frame kinds, carried as the first byte of every sealed plaintext.
const (
	// FrameJSON is an envelope on its own — every control message.
	FrameJSON = 0x00
	// FrameBlob is an envelope followed by raw bytes: a file chunk, a camera
	// frame, anything whose payload is bulk data rather than structure.
	FrameBlob = 0x01
)

// MaxFrameSize bounds one sealed frame. It has to clear the largest blob
// (ChunkSize) plus its envelope with room to spare, and it is the only thing
// standing between a hostile peer and an allocation the size of its
// imagination.
const MaxFrameSize = 4 << 20

var errShortFrame = errors.New("protocol: truncated frame")

// EncodeFrame renders an envelope, optionally carrying raw bytes alongside it.
//
// Bulk payloads travel *beside* the JSON rather than inside it because the
// transport is already binary: the sealed frame is written as a WebSocket
// binary message either way. Base64 inside the envelope would inflate every
// byte by a third and cost an encode on one CPU and a decode on the other, to
// end up in the same place.
func EncodeFrame(env *Envelope, blob []byte) ([]byte, error) {
	meta, err := json.Marshal(env)
	if err != nil {
		return nil, err
	}
	if len(blob) == 0 {
		return append([]byte{FrameJSON}, meta...), nil
	}

	frame := make([]byte, 0, 5+len(meta)+len(blob))
	frame = append(frame, FrameBlob)
	frame = binary.BigEndian.AppendUint32(frame, uint32(len(meta)))
	frame = append(frame, meta...)
	frame = append(frame, blob...)
	return frame, nil
}

// DecodeFrame splits a plaintext frame back into its envelope and its bytes.
// The returned blob aliases the input, which is fine for every current caller
// — the chunk is written to disk before the buffer is reused — and is what
// keeps a large transfer from copying every byte twice.
func DecodeFrame(frame []byte) (*Envelope, []byte, error) {
	if len(frame) == 0 {
		return nil, nil, errShortFrame
	}

	switch frame[0] {
	case FrameJSON:
		env := &Envelope{}
		if err := json.Unmarshal(frame[1:], env); err != nil {
			return nil, nil, err
		}
		return env, nil, nil

	case FrameBlob:
		if len(frame) < 5 {
			return nil, nil, errShortFrame
		}
		metaLen := int(binary.BigEndian.Uint32(frame[1:5]))
		// Checked against the frame rather than a constant: a length that
		// overruns the buffer is the one input that turns a slice into a panic.
		if metaLen < 0 || 5+metaLen > len(frame) {
			return nil, nil, errShortFrame
		}
		env := &Envelope{}
		if err := json.Unmarshal(frame[5:5+metaLen], env); err != nil {
			return nil, nil, err
		}
		return env, frame[5+metaLen:], nil
	}

	return nil, nil, fmt.Errorf("protocol: unknown frame kind 0x%02x", frame[0])
}

File: mobile/app/src/main/java/com/switchboard/app/net/Frame.kt
package com.switchboard.app.net

import kotlinx.serialization.json.JsonElement

/**
 * Frame codec, mirroring `backend/internal/protocol/frame.go` byte for byte.
 *
 * Bulk payloads — file chunks, camera frames — travel *beside* the JSON
 * envelope rather than base64 inside it. The sealed frame is written as a
 * WebSocket binary message either way, so encoding them would inflate every
 * byte by a third and buy an encode on this device's CPU and a decode on the
 * desktop's to arrive at the same place.
 *
 * Layout:
 * ```
 * 0x00 | envelope JSON                                  (control frame)
 * 0x01 | uint32 big-endian metaLen | envelope JSON | raw bytes
 * ```
 */
object Frame {
    const val KIND_JSON: Byte = 0x00
    const val KIND_BLOB: Byte = 0x01

    /** One decoded frame: the envelope, and the bytes that rode with it. */
    data class Decoded(val envelope: Envelope, val blob: ByteArray?) {
        // Generated equals/hashCode would compare the ByteArray by identity,
        // which is a trap rather than a feature. Nothing compares these.
        override fun equals(other: Any?) = this === other
        override fun hashCode() = System.identityHashCode(this)
    }

    fun encode(envelope: Envelope, blob: ByteArray? = null): ByteArray {
        val meta = SwitchboardJson
            .encodeToString(Envelope.serializer(), envelope)
            .encodeToByteArray()

        if (blob == null || blob.isEmpty()) {
            return ByteArray(1 + meta.size).also {
                it[0] = KIND_JSON
                meta.copyInto(it, 1)
            }
        }

        return ByteArray(5 + meta.size + blob.size).also {
            it[0] = KIND_BLOB
            it[1] = (meta.size ushr 24).toByte()
            it[2] = (meta.size ushr 16).toByte()
            it[3] = (meta.size ushr 8).toByte()
            it[4] = meta.size.toByte()
            meta.copyInto(it, 5)
            blob.copyInto(it, 5 + meta.size)
        }
    }

    /** Throws on a malformed frame; the caller drops it. */
    fun decode(frame: ByteArray): Decoded {
        require(frame.isNotEmpty()) { "empty frame" }

        return when (frame[0]) {
            KIND_JSON -> Decoded(parse(frame, 1, frame.size), null)

            KIND_BLOB -> {
                require(frame.size >= 5) { "truncated blob header" }
                // Read as a Long: a hostile length with the high bit set would
                // otherwise arrive as a negative Int and slip past the bounds
                // check that is supposed to catch it.
                val metaLen = ((frame[1].toLong() and 0xFF) shl 24) or
                    ((frame[2].toLong() and 0xFF) shl 16) or
                    ((frame[3].toLong() and 0xFF) shl 8) or
                    (frame[4].toLong() and 0xFF)
                require(5 + metaLen <= frame.size) { "blob metadata overruns the frame" }
                val end = (5 + metaLen).toInt()
                Decoded(parse(frame, 5, end), frame.copyOfRange(end, frame.size))
            }

            else -> throw IllegalArgumentException("unknown frame kind ${frame[0]}")
        }
    }

    private fun parse(frame: ByteArray, from: Int, to: Int): Envelope =
        SwitchboardJson.decodeFromString(
            Envelope.serializer(),
            frame.decodeToString(from, to)
        )
}

/** A control frame's payload, or null when it carried none. */
val Frame.Decoded.payload: JsonElement? get() = envelope.payload
